app/models/container_deployment.rb

```diff
@@ -6,5 +6,314 @@ class ContainerDeployment < ApplicationRecord
+    template = <<eos
+[OSEv3:children]
+nodes
+#{"masters" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+
+[OSEv3:vars]
+ansible_ssh_user=#{ssh_auth.userid}
+deployment_type=openshift-enterprise
+rhsub_user=#{rhsm_auth.userid}
+rhsub_pass=#{rhsm_auth.password_encrypted}
+rhsub_pool=#{rhsm_auth.rhsm_sku}
+ansible_sudo=true
+
+[nodes]
+#{(roles_addresses("node") - [roles_addresses("deployment_master")]).join("\n") unless roles_addresses("node").empty?}
+
+#{"[masters]" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+#{(roles_addresses("master") - [roles_addresses("deployment_master")]).join("\n") unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+eos
+    template
+  end
+
+  def generate_roles
+    result = {}
+    roles = []
+    container_deployment_nodes.each do |node|
+      roles << node.roles
+    end
+    roles = roles.flatten.uniq
+    result["master"] = {"osm_use_cockpit" => "false"} if roles.include? "master"
+    result["node"] = {} if roles.include? "node"
+    result["storage"] = {} if roles.include? "storage"
+    result
+  end
+
+  def generate_ansible_yaml
+    hash = {}
+    hash["ansible_config"] = ANSIBLE_CONFIG_LOCATION
+    hash["ansible_log_path"] = ANSIBLE_CONFIG_LOG
+    hash["ansible_inventory_path"] = ANSIBLE_CONFIG_INVENTORY_PATH
+    hash["ansible_ssh_user"] = ssh_auth.userid
+    hash["deployment"] = {}
+    hash["deployment"]["hosts"] = container_deployment_nodes.collect(&:to_ansible_config_format)
+    hash["deployment"]["roles"] = generate_roles
+    hash["version"] = version
+    hash["variant_version"] = kind.include?("origin") ? '1.2' : '3.2'
+    hash["variant"] = kind
+    # hash[:deployment][:roles][:master][:openshift_master_identity_providers] = []
+    # hash[:deployment][:roles][:master][:openshift_master_identity_providers] << identity_provider_auth.first.send(:to_ansible_config_format)
+    unless identity_provider_auth.first.htpassd_users.empty?
+      hash[:deployment][:roles][:master][:openshift_master_htpasswd_users] = {}
+      hash[:deployment][:roles][:master][:openshift_master_htpasswd_users] = identity_provider_auth.first.htpassd_users
+    end
+    hash.to_yaml.gsub("---\n", '')
+  end
+
+  def extract_public_ip_or_hostname(deployment_node)
+    return deployment_node.address if deployment_node.vm_id.nil?
+    vm = deployment_node.vm
+    hostname = vm.hostnames
+    if hostname.empty?
+      ips = vm.hardware.ipaddresses
+      ips.each do |ip|
+        return ip if public_ip? ip
+      end
+    end
+    hostname.first || ips.last
+  end
+
+  def public_ip?(ip)
+    require "ipaddr"
+    private = false
+    ip_addr = IPAddr.new(ip).to_i
+    ranges = ['10.0.0.0|10.255.255.255',
+              '172.16.0.0|172.31.255.255',
+              '192.168.0.0|192.168.255.255',
+              '169.254.0.0|169.254.255.255',
+              '127.0.0.0|127.255.255.255']
```

Comment:
This does not make sense to create a string that you later need to parse...you can just use normal arrays here.

```ruby
    ranges = [
      ['10.0.0.0',    '10.255.255.255'],
      ['172.16.0.0',  '172.31.255.255'],
      ['192.168.0.0', '192.168.255.255'],
      ['169.254.0.0', '169.254.255.255'],
      ['127.0.0.0',   '127.255.255.255']
    ]
    ranges.each do |low, high|
      ...
```

Alternately,

```ruby
    ranges = [
      '10.0.0.0',    '10.255.255.255',
      '172.16.0.0',  '172.31.255.255',
      '192.168.0.0', '192.168.255.255',
      '169.254.0.0', '169.254.255.255',
      '127.0.0.0',   '127.255.255.255'
    ].collect { |ip| IPAddr.new(ip).to_i }

    private = ranges.each_pair(2).any? { |high, low| (low..high).cover?(ip_addr) }
```